cmd/bench.go

```
@@ -232,32 +262,278 @@ func (r *goBenchRunner) run(ctx context.Context, ectx *evalContext, params bench
 		}
 
 		if params.metrics {
-			// For each histogram add their values to the benchmark results.
-			// Note: If there are many metrics this gets super verbose.
-			for histName, metric := range hist.All() {
-				histValues, ok := metric.(map[string]interface{})
-				if !ok {
-					continue
-				}
-				for metricName, rawValue := range histValues {
-					unit := fmt.Sprintf("%s_%s", histName, metricName)
+			reportMetrics(b, hist.All())
+		}
+	})
+
+	return br, benchErr
+}
+
+func benchE2E(ctx context.Context, args []string, params benchmarkCommandParams, w io.Writer) error {
+	host := "localhost"
+	port := "18181"
+
+	logger := logging.New()
+	logger.SetLevel(logging.Error)
+
+	paths := params.dataPaths.v
+	if len(params.bundlePaths.v) > 0 {
+		paths = append(paths, params.bundlePaths.v...)
+	}
+
+	rtParams := runtime.Params{
+		Addrs:                  &[]string{fmt.Sprintf("%s:%s", host, port)},
+		Paths:                  paths,
+		Logger:                 logger,
+		BundleMode:             params.bundlePaths.isFlagSet(),
+		SkipBundleVerification: true,
+		EnableVersionCheck:     false,
+		GracefulShutdownPeriod: params.gracefulShutdownPeriod,
+		ShutdownWaitPeriod:     params.shutdownWaitPeriod,
+	}
+
+	rt, err := runtime.NewRuntime(ctx, rtParams)
+	if err != nil {
+		return err
+	}
+
+	cctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	initChannel := rt.Manager.ServerInitializedChannel()
+	go func() {
+		if err := rt.Serve(cctx); err != nil {
+			os.Exit(1)
```

Exit here is brutal. Can we return a failure so the caller can do b.Fail()?